src/cirrus/lib2/events.py

```diff
@@ -0,0 +1,249 @@
+import json
+import os
+from contextlib import AbstractContextManager, contextmanager
+from datetime import datetime, timezone
+from logging import Logger, getLogger
+from typing import Dict
+
+import boto3
+
+from .enums import StateEnum, WFEventType
+from .eventdb import EventDB
+from .statedb import StateDB
+from .utils import SNSPublisher
+
+logger = getLogger(__name__)
+
+
+class WorkflowEventManager:
+    """A class for managing payload state change events, including storage of
+    state (DynamoDB), activity (Timestream), and notifications of decisions and/or status
+    changes (SNS)."""
+
+    def __init__(
+        self,
+        logger: Logger = logger,
+        boto3_session: boto3.Session = None,
+        statedb: StateDB = None,
+        eventdb: EventDB = None,
+        batch_size: int = 10,
+    ):
+        self.logger = logger
+        self._boto3_session = boto3_session if boto3_session else boto3.Session()
+        wf_event_topic_arn = os.getenv("CIRRUS_WORKFLOW_EVENT_TOPIC_ARN")
+        self.event_publisher = (
+            SNSPublisher(
+                wf_event_topic_arn,
+                logger=logger,
+                boto3_session=self._boto3_session,
+                batch_size=batch_size,
+            )
+            if wf_event_topic_arn
+            else None
+        )
+        self.statedb = (
+            statedb if statedb is not None else StateDB(session=self._boto3_session)
+        )
+        self.eventdb = (
+            eventdb if eventdb is not None else EventDB(session=self._boto3_session)
+        )
+
+    def flush(self):
+        if self.event_publisher:
+            self.event_publisher.execute()
+
+    @classmethod
+    @contextmanager
+    def handler(
+        cls: "WorkflowEventManager", *args, **kwargs
+    ) -> AbstractContextManager["WorkflowEventManager"]:
+        wfem = cls(*args, **kwargs)
+        try:
+            yield wfem
+        finally:
+            wfem.flush()
+
+    def announce(
+        self,
+        event_type: WFEventType,
+        payload: dict = None,
+        payload_id: str = None,
+        extra_message: dict = None,
+    ) -> None:
+        """
+        Construct message payload and publish to WorkflowEventTopic.
+
+        Args:
+            event_type (WFEventType): the type of event which occurred
+            payload (dict): The ProcessPayload which triggered the event.
+            payload_id (str): The ID of the payload.  Must match payload["id"] if that
+                              exists / is  provided.
+            extra_message (dict): dictionary of additional items to be placed into the
+                                  SNS message body.
+        """
+        if self.event_publisher is None:
+            return
+
+        if payload is None:
+            if payload_id is None:
+                raise ValueError("must specify payload_id or payload")
```

Comment:
It seems like you could collapse these conditions:

```suggestion
        if payload is None and payload_id is None:
            raise ValueError("must specify payload_id or payload")
```